src/doc/writer.rs

```diff
@@ -0,0 +1,259 @@
+use crate::doc::doc::*;
+use crate::error::{Error, FileIOAction, FileKind};
+use handlebars::Handlebars;
+use pulldown_cmark;
+use serde::Serialize;
+use std::collections::HashMap;
+use std::fs::{read_to_string, File};
+use std::io::Write;
+use std::path::PathBuf;
+
+#[derive(Debug, Serialize, Clone)]
+struct ModuleItem {
+    pub name: String,
+    pub doc: Option<String>,
+    pub signature: String,
+}
+
+#[derive(Debug, Serialize, Clone)]
+struct Module {
+    pub module_name: String,
+    pub project_name: String,
+    pub project_version: String,
+    pub all_modules: Option<Vec<Module>>,
+    pub module_doc: String,
+    pub module_src: String,
+    pub types_doc: Vec<ModuleItem>,
+    pub functions_doc: Vec<ModuleItem>,
+    pub readme: String,
+}
+
+pub struct DocWriter<'a> {
+    project_name: String,
+    project_version: String,
+    doc_dir: PathBuf,
+    modules: Vec<Module>,
+    registry: Handlebars<'a>,
+    package_root: PathBuf,
+}
+
+fn render_markdown(text: String) -> String {
+    let mut s = String::with_capacity(text.len() * 3 / 2);
+    let p = pulldown_cmark::Parser::new(&*text);
+    pulldown_cmark::html::push_html(&mut s, p);
+    s
+}
+
+handlebars_helper!(markdown: |text: str| render_markdown(text.to_string()));
+
+impl DocWriter<'_> {
+    pub fn new(
+        project_name: String,
+        project_version: String,
+        doc_dir: PathBuf,
+        package_root: PathBuf,
+    ) -> Self {
+        DocWriter {
+            project_name,
+            project_version,
+            doc_dir,
+            registry: {
+                let module_template = std::include_str!("views/module.hbs");
+                let index_template = std::include_str!("views/index.hbs");
+                let mut registry = Handlebars::new();
+                registry.register_helper("markdown", Box::new(markdown));
+                registry
+                    .register_template_string("module", module_template)
+                    .unwrap_or_else(|_e| panic!("Doc Template is missing"));
+                registry
+                    .register_template_string("index", index_template)
+                    .unwrap_or_else(|_e| panic!("Doc Template is missing"));
+                registry
+            },
+            modules: vec![],
+            package_root,
+        }
+    }
+
+    pub fn add_chunk(self: &mut Self, chunk: EEP48DocChunk) {
+        if chunk.anno.file.ends_with("_test") {
+            return;
+        }
+        self.modules.push(self.chunk_to_module(chunk));
+    }
+
+    fn write_index(self: &Self, module: &Module) -> Result<(), Error> {
+        let rendered = self
+            .registry
+            .render("index", module)
+            .map_err(|_| Error::FileIO {
+                path: self.doc_dir.join("index.html"),
+                action: FileIOAction::Create,
```

Comment:
It's a parse error, so `Parse`.

Though looking at the code it could be several other errors here. We can make this more precise in a later PR